flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/HighAvailabilityServicesUtils.java

```diff
@@ -82,6 +80,31 @@ public static HighAvailabilityServices createAvailableOrEmbeddedServices(
         }
     }
 
+    @Nonnull
+    private static HighAvailabilityServices createZooKeeperHaServices(
```

Comment:
Why is there a `HighAvailabilityServicesFactory` implementation for k8s but not for ZooKeeper?

Reply:
Because you can only activate the K8s ha services by configuring a factory. This has historical reasons.